File: Sources/SwiftMTHColor/ColorSubstitution.swift
/// Substitute MUD color codes in the input string, producing ANSI escape sequences
/// appropriate for the given color depth.
///
/// Supports three color code syntaxes:
/// - `^a`-`^z` / `^A`-`^Z`: 32 named MUD colors (dark/bright)
/// - `<F000>`-`<FFFF>`: True color foreground (hex RGB)
/// - `<B000>`-`<BFFF>`: True color background (hex RGB)
/// - `^^`: Escaped literal caret
/// - `^?`: Random color
public func substituteColor(_ input: String, depth: ColorDepth) -> String {
    var bytes = Array(input.utf8)
    bytes.append(0) // null terminator for C-like iteration
    var output: [UInt8] = []
    output.reserveCapacity(bytes.count * 6)
    substituteColorBytes(bytes, into: &output, colors: depth.rawValue)
    return String(decoding: output, as: UTF8.self)
}

/// Internal recursive implementation operating on byte arrays.
/// Returns number of bytes written to output (for recursive calls).
@discardableResult
private func substituteColorBytes(_ input: [UInt8], into output: inout [UInt8], colors: Int) -> Int {
    let startCount = output.count
    var oldF: [UInt8] = [0, 0, 0, 0, 0, 0, 0] // tracks last foreground code
    var oldB: [UInt8] = [0, 0, 0, 0, 0, 0, 0] // tracks last background code
    var i = 0

    while i < input.count && input[i] != 0 {
        switch input[i] {
        case UInt8(ascii: "^"):
            let next = i + 1 < input.count ? input[i + 1] : 0
            if ColorTables.is32c(next) != 0 {
                // Skip pattern: ^r^^g skips ^r, processes ^g
                if i + 3 < input.count && input[i + 2] == UInt8(ascii: "^") && ColorTables.is32c(input[i + 3]) != 0 {
                    i += 2
                    continue
                }

                if colors != 0 {
                    if next == UInt8(ascii: "?") {
                        // Random color — generate a random <Fxxx> code and recurse
                        let rndCode = randomTrueColorCode()
                        var rndBytes = Array(rndCode.utf8)
                        rndBytes.append(0)
                        substituteColorBytes(rndBytes, into: &output, colors: colors)
                    } else if oldF[0] != input[i] || oldF[1] != next {
                        // Different from last foreground — emit ANSI
                        if next >= UInt8(ascii: "a") && next <= UInt8(ascii: "z") {
                            let idx = Int(next) - Int(UInt8(ascii: "a"))
                            let expanded = ColorTables.alphabetFgcDark[idx]
                            var expandedBytes = Array(expanded.utf8)
                            expandedBytes.append(0)
                            let cappedColors = colors < 256 ? colors : 256
                            substituteColorBytes(expandedBytes, into: &output, colors: cappedColors)
                        } else {
                            let idx = Int(next) - Int(UInt8(ascii: "A"))
                            let expanded = ColorTables.alphabetFgcBold[idx]
                            var expandedBytes = Array(expanded.utf8)
                            expandedBytes.append(0)
                            let cappedColors = colors < 256 ? colors : 256
                            substituteColorBytes(expandedBytes, into: &output, colors: cappedColors)
                        }
                    }
                }
                // Update old_f and advance past 2-char code
                oldF[0] = input[i]
                oldF[1] = next
                i += 2

            } else {
                // Not a valid 32-color code
                if next == UInt8(ascii: "^") {
                    // ^^ escape — skip first ^, output second
                    i += 1
                }
                output.append(input[i])
                i += 1
            }

        case UInt8(ascii: "<"):
            if matchesForegroundCode(input, at: i) {
                let c2 = input[i + 2], c3 = input[i + 3], c4 = input[i + 4]
                let normalized = normalizedFCode(c2, c3, c4)

                if !caseInsensitiveMatch6(oldF, input, at: i) && colors != 0 {
                    if colors == 4096 {
                        let r = ColorTables.tcVal(c2)
                        let g = ColorTables.tcVal(c3)
                        let b = ColorTables.tcVal(c4)
                        appendString(&output, "\u{1B}[38;2;\(r);\(g);\(b)m")
                    } else if colors == 256 {
                        let idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                        appendString(&output, "\u{1B}[38;5;\(idx)m")
                    } else {
                        // 16 colors — recurse through ANSI table
                        let idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                        let ansi = ColorTables.ansiForeground[idx]
                        var ansiBytes = Array(ansi.utf8)
                        ansiBytes.append(0)
                        substituteColorBytes(ansiBytes, into: &output, colors: colors)
                    }
                }
                // Update old_f to normalized form and advance
                for j in 0..<6 { oldF[j] = normalized[j] }
                i += 6

            } else if matchesBackgroundCode(input, at: i) {
                let c2 = input[i + 2], c3 = input[i + 3], c4 = input[i + 4]
                let normalized = normalizedFCode(c2, c3, c4) // C uses "<F%c%c%c>" for both

                if !caseInsensitiveMatch6(oldB, input, at: i) && colors != 0 {
                    if colors == 4096 {
                        let r = ColorTables.tcVal(c2)
                        let g = ColorTables.tcVal(c3)
                        let b = ColorTables.tcVal(c4)
                        appendString(&output, "\u{1B}[48;2;\(r);\(g);\(b)m")
                    } else if colors == 256 {
                        let idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                        appendString(&output, "\u{1B}[48;5;\(idx)m")
                    } else {
                        let idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                        let ansi = ColorTables.ansiBackground[idx]
                        var ansiBytes = Array(ansi.utf8)
                        ansiBytes.append(0)
                        substituteColorBytes(ansiBytes, into: &output, colors: colors)
                    }
                }
                // C stores as "<F%c%c%c>" for old_b too (yes, F not B — matches C bug/behavior)
                for j in 0..<6 { oldB[j] = normalized[j] }
                i += 6

            } else {
                output.append(input[i])
                i += 1
            }

        default:
            output.append(input[i])
            i += 1
        }
    }

    return output.count - startCount
}

// MARK: - Helpers

private func appendString(_ output: inout [UInt8], _ s: String) {
    output.append(contentsOf: s.utf8)
}

/// Check if input at position i matches <Fxxx> or <fxxx> pattern.
private func matchesForegroundCode(_ input: [UInt8], at i: Int) -> Bool {
    guard i + 5 < input.count else { return false }
    let ch1 = input[i + 1]
    return (ch1 == UInt8(ascii: "F") || ch1 == UInt8(ascii: "f"))
        && isHexDigit(input[i + 2])
        && isHexDigit(input[i + 3])
        && isHexDigit(input[i + 4])
        && input[i + 5] == UInt8(ascii: ">")
}

/// Check if input at position i matches <Bxxx> or <bxxx> pattern.
private func matchesBackgroundCode(_ input: [UInt8], at i: Int) -> Bool {
    guard i + 5 < input.count else { return false }
    let ch1 = input[i + 1]
    return (ch1 == UInt8(ascii: "B") || ch1 == UInt8(ascii: "b"))
        && isHexDigit(input[i + 2])
        && isHexDigit(input[i + 3])
        && isHexDigit(input[i + 4])
        && input[i + 5] == UInt8(ascii: ">")
}

private func isHexDigit(_ c: UInt8) -> Bool {
    (c >= UInt8(ascii: "0") && c <= UInt8(ascii: "9"))
        || (c >= UInt8(ascii: "a") && c <= UInt8(ascii: "f"))
        || (c >= UInt8(ascii: "A") && c <= UInt8(ascii: "F"))
}

private func toUpper(_ c: UInt8) -> UInt8 {
    if c >= UInt8(ascii: "a") && c <= UInt8(ascii: "z") {
        return c - 32
    }
    return c
}

/// Produce the normalized "<Fxxx>" form used for old_f/old_b tracking.
/// The C code uses `sprintf(old_f, "<F%c%c%c>", pti[2], pti[3], pti[4])`.
private func normalizedFCode(_ c2: UInt8, _ c3: UInt8, _ c4: UInt8) -> [UInt8] {
    [UInt8(ascii: "<"), UInt8(ascii: "F"), c2, c3, c4, UInt8(ascii: ">"), 0]
}

/// Case-insensitive comparison of first 6 bytes of `old` against `input[at..<at+6]`.
private func caseInsensitiveMatch6(_ old: [UInt8], _ input: [UInt8], at i: Int) -> Bool {
    guard i + 5 < input.count else { return false }
    for j in 0..<6 {
        if toUpper(old[j]) != toUpper(input[i + j]) {
            return false
        }
    }
    return true
}

/// Generate a random true color code string like "<F8A3>".
private func randomTrueColorCode() -> String {
    let hex = ColorTables.decToHex
    let r = hex[Int.random(in: 0..<16)]
    let g = hex[Int.random(in: 0..<16)]
    let b = hex[Int.random(in: 0..<16)]
    return "<F\(r)\(g)\(b)>"
}

File: kotlin/mth-color/src/main/kotlin/mth/color/ColorSubstitution.kt
package mth.color

import kotlin.random.Random

fun substituteColor(input: String, depth: ColorDepth): String {
    val bytes = input.toByteArray(Charsets.UTF_8) + 0.toByte() // null terminator
    val output = mutableListOf<Byte>()
    substituteColorBytes(bytes, output, depth.colors)
    return String(output.toByteArray(), Charsets.UTF_8)
}

private fun substituteColorBytes(input: ByteArray, output: MutableList<Byte>, colors: Int): Int {
    val startCount = output.size
    val oldF = ByteArray(7) // tracks last foreground code
    val oldB = ByteArray(7) // tracks last background code
    var i = 0

    while (i < input.size && input[i] != 0.toByte()) {
        val c = input[i].toInt() and 0xFF

        when (c) {
            '^'.code -> {
                val next = if (i + 1 < input.size) input[i + 1] else 0.toByte()
                if (ColorTables.is32c(next) != 0) {
                    // Skip pattern: ^r^^g skips ^r, processes ^g
                    if (i + 3 < input.size &&
                        input[i + 2] == '^'.code.toByte() &&
                        ColorTables.is32c(input[i + 3]) != 0
                    ) {
                        i += 2
                        continue
                    }

                    if (colors != 0) {
                        val nextUnsigned = next.toInt() and 0xFF
                        if (nextUnsigned == '?'.code) {
                            // Random color
                            val rndCode = randomTrueColorCode()
                            val rndBytes = rndCode.toByteArray(Charsets.UTF_8) + 0.toByte()
                            substituteColorBytes(rndBytes, output, colors)
                        } else if (oldF[0] != input[i] || oldF[1] != next) {
                            // Different from last foreground -- emit ANSI
                            if (nextUnsigned in 'a'.code..'z'.code) {
                                val idx = nextUnsigned - 'a'.code
                                val expanded = ColorTables.alphabetFgcDark[idx]
                                val expandedBytes = expanded.toByteArray(Charsets.UTF_8) + 0.toByte()
                                val cappedColors = if (colors < 256) colors else 256
                                substituteColorBytes(expandedBytes, output, cappedColors)
                            } else {
                                val idx = nextUnsigned - 'A'.code
                                val expanded = ColorTables.alphabetFgcBold[idx]
                                val expandedBytes = expanded.toByteArray(Charsets.UTF_8) + 0.toByte()
                                val cappedColors = if (colors < 256) colors else 256
                                substituteColorBytes(expandedBytes, output, cappedColors)
                            }
                        }
                    }
                    // Update oldF
                    oldF[0] = input[i]
                    oldF[1] = next
                    i += 2
                } else {
                    if ((next.toInt() and 0xFF) == '^'.code) {
                        // ^^ escape -- skip first ^, output second
                        i += 1
                    }
                    output.add(input[i])
                    i += 1
                }
            }

            '<'.code -> {
                if (matchesForegroundCode(input, i)) {
                    val c2 = input[i + 2]
                    val c3 = input[i + 3]
                    val c4 = input[i + 4]
                    val normalized = normalizedFCode(c2, c3, c4)

                    if (!caseInsensitiveMatch6(oldF, input, i) && colors != 0) {
                        if (colors == 4096) {
                            val r = ColorTables.tcVal(c2)
                            val g = ColorTables.tcVal(c3)
                            val b = ColorTables.tcVal(c4)
                            appendString(output, "\u001B[38;2;${r};${g};${b}m")
                        } else if (colors == 256) {
                            val idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                            appendString(output, "\u001B[38;5;${idx}m")
                        } else {
                            // 16 colors -- recurse through ANSI table
                            val idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                            val ansi = ColorTables.ansiForeground[idx]
                            val ansiBytes = ansi.toByteArray(Charsets.UTF_8) + 0.toByte()
                            substituteColorBytes(ansiBytes, output, colors)
                        }
                    }
                    for (j in 0 until 6) oldF[j] = normalized[j]
                    i += 6
                } else if (matchesBackgroundCode(input, i)) {
                    val c2 = input[i + 2]
                    val c3 = input[i + 3]
                    val c4 = input[i + 4]
                    val normalized = normalizedFCode(c2, c3, c4) // C uses F for both

                    if (!caseInsensitiveMatch6(oldB, input, i) && colors != 0) {
                        if (colors == 4096) {
                            val r = ColorTables.tcVal(c2)
                            val g = ColorTables.tcVal(c3)
                            val b = ColorTables.tcVal(c4)
                            appendString(output, "\u001B[48;2;${r};${g};${b}m")
                        } else if (colors == 256) {
                            val idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                            appendString(output, "\u001B[48;5;${idx}m")
                        } else {
                            val idx = 16 + ColorTables.x256cVal(c2) * 36 + ColorTables.x256cVal(c3) * 6 + ColorTables.x256cVal(c4)
                            val ansi = ColorTables.ansiBackground[idx]
                            val ansiBytes = ansi.toByteArray(Charsets.UTF_8) + 0.toByte()
                            substituteColorBytes(ansiBytes, output, colors)
                        }
                    }
                    for (j in 0 until 6) oldB[j] = normalized[j]
                    i += 6
                } else {
                    output.add(input[i])
                    i += 1
                }
            }

            else -> {
                output.add(input[i])
                i += 1
            }
        }
    }

    return output.size - startCount
}

private fun isHexDigit(c: Byte): Boolean {
    val v = c.toInt() and 0xFF
    return v in '0'.code..'9'.code || v in 'A'.code..'F'.code || v in 'a'.code..'f'.code
}

private fun toUpper(c: Byte): Byte {
    val v = c.toInt() and 0xFF
    return if (v in 'a'.code..'z'.code) (v - 32).toByte() else c
}

private fun matchesForegroundCode(input: ByteArray, at: Int): Boolean {
    if (at + 5 >= input.size) return false
    if ((input[at].toInt() and 0xFF) != '<'.code) return false
    val tag = input[at + 1].toInt() and 0xFF
    if (tag != 'F'.code && tag != 'f'.code) return false
    if (!isHexDigit(input[at + 2])) return false
    if (!isHexDigit(input[at + 3])) return false
    if (!isHexDigit(input[at + 4])) return false
    if ((input[at + 5].toInt() and 0xFF) != '>'.code) return false
    return true
}

private fun matchesBackgroundCode(input: ByteArray, at: Int): Boolean {
    if (at + 5 >= input.size) return false
    if ((input[at].toInt() and 0xFF) != '<'.code) return false
    val tag = input[at + 1].toInt() and 0xFF
    if (tag != 'B'.code && tag != 'b'.code) return false
    if (!isHexDigit(input[at + 2])) return false
    if (!isHexDigit(input[at + 3])) return false
    if (!isHexDigit(input[at + 4])) return false
    if ((input[at + 5].toInt() and 0xFF) != '>'.code) return false
    return true
}

private fun normalizedFCode(c2: Byte, c3: Byte, c4: Byte): ByteArray {
    return byteArrayOf(
        '<'.code.toByte(),
        'F'.code.toByte(),
        toUpper(c2),
        toUpper(c3),
        toUpper(c4),
        '>'.code.toByte(),
        0.toByte()
    )
}

private fun caseInsensitiveMatch6(old: ByteArray, input: ByteArray, at: Int): Boolean {
    for (j in 0 until 6) {
        if (at + j >= input.size) return false
        if (toUpper(old[j]) != toUpper(input[at + j])) return false
    }
    return true
}

private fun appendString(output: MutableList<Byte>, s: String) {
    for (b in s.toByteArray(Charsets.UTF_8)) {
        output.add(b)
    }
}

private fun randomTrueColorCode(): String {
    val r = ColorTables.decToHex[Random.nextInt(16)]
    val g = ColorTables.decToHex[Random.nextInt(16)]
    val b = ColorTables.decToHex[Random.nextInt(16)]
    return "<F${r}${g}${b}>"
}
